### PR TITLE
snapstate: support getting new bases/default-providers on refresh

### PR DESCRIPTION
**Pull request by mvo5**
The current code does not pull in changed bases or changed
default-providers on refreshes. This PR adds support to do this.

This is an early preview, I work on spread tests next.

### CODE REVIEW

File: overlord/snapstate/snapstate_test.go [hunk 1]
```diff
@@ -2188,6 +2188,172 @@ func (s *snapmgrTestSuite) TestUpdateRunThrough(c *C) {
...
+	s.state.Unlock()
+	defer s.snapmgr.Stop()
+	s.settle(c)
+	s.state.Lock()
+
+	c.Check(s.fakeStore.downloads, DeepEquals, []fakeDownload{
+		{macaroon: s.user.StoreMacaroon, name: "snap-content-plug"},
+	})
+}
+
```
Comment (stolowski) [inline overlord/snapstate/snapstate_test.go]: Thank you for extensive tests :heart:

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -781,6 +781,8 @@ func doUpdate(st *state.State, names []string, updates []*snap.Info, params func
 		}
 
 		snapsup := &SnapSetup{
+			Base:         update.Base,
+			Prereq:       defaultContentPlugProviders(st, update),
```
Comment (stolowski) [inline overlord/snapstate/snapstate.go]: I like the fact how it extends existing funcionality with just this simple change :+1:

Comment (bboozzoo) [reply overlord/snapstate/snapstate.go]: Very nice how the integration bits fall into place.

**Comment by pedronis** [comment]
marking blocked because missing spread tests

File: tests/main/base-snaps/task.yaml [hunk 3]
```diff
@@ -5,24 +5,46 @@ execute: |
     #shellcheck source=tests/lib/snaps.sh
     . $TESTSLIB/snaps.sh
 
-    echo "Ensure a snap that requires a unavailable base snap can not be installed"
-    snap pack "$TESTSLIB"/snaps/test-snapd-requires-base
-    if install_local test-snapd-requires-base; then
+    echo "Ensure a snap that requires an unavailable base snap can not be installed"
+    cp -a $TESTSLIB/snaps/test-snapd-requires-base .
```
Comment (bboozzoo) [inline tests/main/base-snaps/task.yaml]: `"$TESTSLIB"`

File: tests/main/base-snaps/task.yaml [hunk 4]
```diff
@@ -5,24 +5,46 @@ execute: |
...
+    snap list | MATCH test-snapd-base
 
-    echo "Ensure the bare base works"
+    echo "------------------------------------------------------------------"
+
+    echo "Cleanup for the next test"
+    snap remove test-snapd-requires-base test-snapd-base
 
+    echo "First install snap without base"
+    cp -a $TESTSLIB/snaps/test-snapd-requires-base .
```
Comment (bboozzoo) [inline tests/main/base-snaps/task.yaml]: `"$TESTSLIB"`

File: tests/main/interfaces-content-default-provider/task.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+# This test purges the state which causes the device to reinitialize
+# with (potentially) a different core snap. Running this on core will
+# also trigger a reboot in the middle of the tests because the new
+# core will be applied. So skip the test on core devices.
+systems: [-ubuntu-core-16-*]
+
+execute: |
+    echo "A snap with no default-provider is installed"
+    cp -a $TESTSLIB/snaps/test-snapd-content-plug .
```
Comment (bboozzoo) [inline tests/main/interfaces-content-default-provider/task.yaml]: `"$TESTSLIB"`

**Pull request merged**